docs/spec/distributing.rst

```diff
@@ -10,72 +10,428 @@ Stub files
+constructs not decribed here, and tool authors are encouraged to experiment with
+additional features.
+
+.. _stub-file-syntax:
+
+Syntax
+^^^^^^
+
+Stub files are syntactically valid Python files in the earliest Python version
+that is not yet end-of-life. They use a ``.pyi`` suffix. The Python syntax used
+by stub files is independent from the Python versions supported by the
+implementation, and from the Python version the type checker runs under (if
+any). Therefore, stub authors should use the latest syntax features available in
+the earliest supported version, even if the implementation supports older
+versions. Type checker authors are encouraged to support syntax features from
+newer versions, although stub authors should not use such features if they wish
+to maintain compatibility with all type checkers.
+
+For example, Python 3.7 added the ``async`` keyword (see :pep:`492`).
+Stub authors should use it to mark coroutines, even if the implementation
+still uses the ``@coroutine`` decorator. On the other hand, stubs should not use
+the ``type`` soft keyword from :pep:`695`, introduced in Python 3.12, util
+Python 3.11 reaches end-of-life in October 2027.
+
+Stubs are treated as if ``from __future__ import annotations`` is enabled. In
+particular, built-in generics, pipe union syntax (``X | Y``), and forward
+references can be used.
+
+Type checkers should have a configurable search path for stub files. If a stub
+file is found, the type checker should not read the corresponding "real" module.
+See :ref:`mro` for more information.
+
+.. _stub-file-supported-constructs:
+
+Supported Constructs
+^^^^^^^^^^^^^^^^^^^^
+
+This section lists constructs that type checkers will accept in stub files. Stub
+authors can safely use these constructs. If a construct is marked as
+"unspecified", type checkers may handle it as they best see fit or report an
+error. Linters should usually flag those constructs. Stub authors should avoid
+using them to ensure compatibility across type checkers.
+
+The `typeshed feature tracker <https://github.com/python/typeshed/labels/project%3A%20feature%20tracker>`_ tracks features from the ``typing`` module that are
+not yet supported by all major type checkers. Unless otherwise noted in this
+tracker, stub files support all features from the ``typing`` module of the
+latest released Python version. If a stub uses typing features from a later
+Python version than what the implementation supports, these features can be
+imported from ``typing_extensions`` instead of ``typing``
+
+For example, a stub could use ``Literal``, introduced in Python 3.8,
+for a library supporting Python 3.7+::
+
+    from typing_extensions import Literal
+
+    def foo(x: Literal[""]) -> int: ...
+
+Comments
+""""""""
+
+Standard Python comments are accepted everywhere Python syntax allows them.
+
+Two kinds of structured comments are accepted:
+
+* A ``# type: X`` comment at the end of a line that defines a variable,
```

Comment:
It's a bit odd to discuss `# type: X` and `# type: ignore` comments in the stubs section (other than perhaps to mention that they are supported in stubs). Describing how they work is redundant, since these are covered in other places within the typing spec. I recommend either removing the "Comments" section or paring it back to simply mention that `# type: X` and `# type: ignore` comments are supported in stubs.